packages/protocol/contracts/L2/CrossChainOwned.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity 0.8.20;
+
+import "lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+
+import "../common/EssentialContract.sol";
+import "../signal/ISignalService.sol";
+
+/// @title CrossChainOwned
+/// @notice This contract's owner lives on another chain who uses signal for transaction approval.
+abstract contract CrossChainOwned is EssentialContract {
+    uint64 public ownerChainId; // slot 1
+    uint64 public nextTxId;
+    uint256[49] private __gap;
+
+    event TransactionExecuted(uint64 indexed txId, bytes32 indexed approvalHash);
+
+    error INVALID_PARAMS();
+    error INVALID_EXECUTOR();
+    error TX_NOT_APPROVED();
+    error TX_REVERTED();
+    error NOT_CALLABLE();
+
+    function executeApprovedTransaction(
+        bytes calldata txdata,
+        bytes calldata proof,
+        address executor
+    )
+        external
+    {
+        if (executor != address(0) && executor != msg.sender) {
+            revert INVALID_EXECUTOR();
+        }
+
+        bytes32 approvalHash = _isTransactionApproved(txdata, proof, executor);
+        if (approvalHash == 0) revert TX_NOT_APPROVED();
+
+        (bool success,) = address(this).call(txdata);
```

Comment:
If we let `owner()` be L1 `timelock`, shall we modify those `onlyOwner` methods in TaikoL2 (e.g. `function withdraw(address token, address to) external onlyOwner`), to make it only executable by the L2 executor here instead of `owner()`?

Reply:
Yes, but we don't override `onlyOwner` modifier, we overide `_checkOwner`. It's done in the code.